lib/srv/db/mongodb/autousers_admin.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mongodb
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	mongooptions "go.mongodb.org/mongo-driver/mongo/options"
+	mongoauth "go.mongodb.org/mongo-driver/x/mongo/driver/auth"
+
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// adminClient defines a subset of functions provided by *mongo.Client plus some
+// helper functions.
+type adminClient interface {
+	// Database returns a database handle based on database name.
+	Database(name string, opts ...*options.DatabaseOptions) *mongo.Database
+	// Disconnect disconnects the client.
+	Disconnect(ctx context.Context) error
+	// ServerVersion returns the server version.
+	ServerVersion(ctx context.Context) (*semver.Version, error)
+}
+
+func (e *Engine) connectAsAdmin(ctx context.Context, sessionCtx *common.Session) (adminClient, error) {
+	if adminClientFnCache == nil {
+		client, err := makeBasicAdminClient(ctx, sessionCtx, e)
+		return client, trace.Wrap(err)
+	}
+
+	shareableClient, err := getShareableAdminClient(ctx, adminClientFnCache, sessionCtx, e, makeBasicAdminClient)
+	return shareableClient, trace.Wrap(err)
+}
+
+func getShareableAdminClient(ctx context.Context, cache *utils.FnCache, sessionCtx *common.Session, e *Engine, makeBasicClient makeBasicAdminClientFunc) (*shareableAdminClient, error) {
+	key := struct {
+		databaseName string
+		databaseURI  string
+		adminUser    string
```

Comment:
Probably there is a corner case during CA rotation where a shared client should be reused but a connection needs to be recreated but this should be really narrow case.

Reply:
Good catch. During a rotation, the server should be restarted with a new configuration so the connection will be dead. The connection could be killed for other reasons as well.

Let me try to add a check on whether the shared client is dead after getting it from the cache. If so, just make a non-shared client.

Reply:
Seems the mongo client connection pool has the ability to reconnect even if i restarted mongo docker. So I will not handle that case.

As for the CA rotation case, users should understand that the connection may have issues during rotation. And it should self-heal after cache expires (one minute). I will leave a comment in the code for this case but not worth it to fix properly.